docs/source/reference/deprecation.rst

```diff
@@ -495,3 +495,46 @@ Schedule
 - In Numba 0.57: Support for CUDA toolkit 10.2 was removed.
 - In Numba 0.58: Support for CC < 5.0 and CUDA toolkits 11.0 and 11.1 will be
   removed.
+
+Deprecation of old-style ``NUMBA_CAPTURED_ERRORS``
+==================================================
+
+The old style of the ``NUMBA_CAPTURED_ERRORS`` environment variable is being 
+deprecated in Numba.
+
+Previously, this variable allowed controlling how Numba handles exceptions 
+during compilation that do not inherit from ``numba.core.errors.NumbaError``. 
+The default "old_style" behavior was to capture and wrap these errors, often 
+obscuring the original exception.
+
+The new "new_style" option treats non-``NumbaError`` exceptions as hard errors, 
+propagating them without capturing. This differentiates compilation errors from 
+unintended exceptions during compilation.
+
+The old style will eventually be removed in favor of the new behavior. Users 
+should migrate to setting ``NUMBA_CAPTURED_ERRORS='new_style'`` to opt-in to the 
+new exception handling. This will become the default in the future.
+
+Recommendations
+---------------
+
+- Projects that depends on Numba should set 
+  ``NUMBA_CAPTURED_ERRORS='new_style'`` for testing to find all places where 
+  non-``NumbaError`` exceptions are raised during compilation.
+- Replace any code that raises non-``NumbaError`` to indicate compilation error.
+  For instance, instead of raising ``TypeError``, raise 
+  ``numba.core.errors.NumbaTypeError``.
```

Comment:
```suggestion
- Modify any code that raises a non-``NumbaError`` to indicate a compilation
  error to raise a subclass of ``NumbaError`` instead. For example, instead of
  raising a ``TypeError``, raise a ``numba.core.errors.NumbaTypeError``.
```